File: Range Queries/Subarray_Sum_Queries.py
import sys,math,cmath,random,os
from heapq import heappush,heappop
from bisect import bisect_right,bisect_left
from collections import Counter,deque,defaultdict
from itertools import permutations,combinations
from io import BytesIO, IOBase
from decimal import Decimal,getcontext

BUFSIZE = 8192
class FastIO(IOBase):
    newlines = 0
    def __init__(self, file):
        self._file = file
        self._fd = file.fileno()
        self.buffer = BytesIO()
        self.writable = "x" in file.mode or "r" not in file.mode
        self.write = self.buffer.write if self.writable else None
    def read(self):
        while True:
            b = os.read(self._fd, max(os.fstat(self._fd).st_size, BUFSIZE))
            if not b:
                break
            ptr = self.buffer.tell()
            self.buffer.seek(0, 2), self.buffer.write(b), self.buffer.seek(ptr)
        self.newlines = 0
        return self.buffer.read()
    def readline(self):
        while self.newlines == 0:
            b = os.read(self._fd, max(os.fstat(self._fd).st_size, BUFSIZE))
            self.newlines = b.count(b"\n") + (not b)
            ptr = self.buffer.tell()
            self.buffer.seek(0, 2), self.buffer.write(b), self.buffer.seek(ptr)
        self.newlines -= 1
        return self.buffer.readline()
    def flush(self):
        if self.writable:
            os.write(self._fd, self.buffer.getvalue())
            self.buffer.truncate(0), self.buffer.seek(0)
class IOWrapper(IOBase):
    def __init__(self, file):
        self.buffer = FastIO(file)
        self.flush = self.buffer.flush
        self.writable = self.buffer.writable
        self.write = lambda s: self.buffer.write(s.encode("ascii"))
        self.read = lambda: self.buffer.read().decode("ascii")
        self.readline = lambda: self.buffer.readline().decode("ascii")
sys.stdin, sys.stdout = IOWrapper(sys.stdin), IOWrapper(sys.stdout)

# functions #
# MOD = 998244353
MOD = 10**9 + 7
RANDOM = random.randrange(1,2**62)
def gcd(a,b):
    while b:
        a,b = b,a%b
    return a
def lcm(a,b):
    return a//gcd(a,b)*b
def w(x):
    return x ^ RANDOM
II = lambda : int(sys.stdin.readline().strip())
LII = lambda : list(map(int, sys.stdin.readline().split()))
MI = lambda x : x(map(int, sys.stdin.readline().split()))
SI = lambda : sys.stdin.readline().strip()
SLI = lambda : list(map(lambda x:ord(x)-97,sys.stdin.readline().strip()))
LII_1 = lambda : list(map(lambda x:int(x)-1, sys.stdin.readline().split()))
LII_C = lambda x : list(map(x, sys.stdin.readline().split()))
MATI = lambda x : [list(map(int, sys.stdin.readline().split())) for _ in range(x)]
##

#String hashing: shclass, fenwick sortedlist: fsortl, Number: numtheory/numrare, SparseTable: SparseTable
#Bucket Sorted list: bsortl, Segment Tree(lp,selfop): SegmentTree, bootstrap: bootstrap, Trie: tries
#binary indexed tree: BIT, Segment Tree(point updates): SegmentPoint, Convex Hull: hull, BitArray: bitarray
#Combinatorics: pnc, Diophantine Equations: dpheq, DSU: DSU, Geometry: Geometry, FFT: fft, XOR_dict: xdict
#Persistent Segment Tree: perseg, Binary Trie: b_trie, HLD: hld, String funcs: sf, Segment Tree(lp): SegmentOther
#Graph1(axtree,bl): graphadv, Graph2(khn,sat): 2sat, Graph3(fltn,bprt): graphflatten, Graph4(ep,tp,fw,bmf): graphoth
#Graph5(djik,bfs,dfs): graph, Graph6(dfsin): dfsin, utils: utils, Persistent DSU: perdsu, Merge Sort Tree: sorttree
#2-D BIT: 2DBIT, MonoDeque: mono, matrix: nummat, SuffixAutomaton: sautomaton, linalg: linalg, SquareRtDecomp: sqrt
#Grapth7(bridges): graph_dmgt, FWHT(^,|,&): fwht, Graph8(centr_decom): graph_decom, DpOptimize(knth,dnc): dpopt
#2D-SegmentTree: seg2d, Rollback/Par DSU: rbdsu, Graphflow(hlpp,dnc): graphflow, DynamicConnectivity: odc
#Treap: treap
#Template : https://github.com/OmAmar106/Template-for-Competetive-Programming
# if os.environ.get('LOCAL'):sys.stdin = open(r'input.txt', 'r');sys.stdout = sys.stderr = open(r'output.txt','w')
# if os.environ.get('LOCAL'):import hashlib;print('Hash Value :',hashlib.md5(open(__file__, 'rb').read()).hexdigest());

class Node:
    def __init__(self,suff,pref,full,maxi):
        self.suff = suff
        self.pref = pref
        self.full = full
        self.maxi = maxi
class SegmentTree:
    def f(node1,node2):
        return Node(max(node2.suff,node1.suff+node2.full),max(node1.pref,node1.full+node2.pref),node1.full+node2.full,max(node1.maxi,node2.maxi,node2.pref+node1.suff))
    def __init__(self,L):
        self.n = len(L)
        self.tree = [None]*((1<<(self.n.bit_length()+1)))
        self.build(1,0,len(L)-1,L)
    def build(self,node,start,end,L):
        st = [(node,start,end,False)]
        while st:
            node,start,end,flag = st.pop()
            if start>end:
                continue
            elif start==end:
                self.tree[node] = Node(L[start],L[start],L[start],L[start])
                continue
            if not flag:
                st.append((node,start,end,True))
                mid = (start+end)>>1
                st.append((2*node,start,mid,False))
                st.append((2*node+1,mid+1,end,False))
            else:
                self.tree[node] = SegmentTree.f(self.tree[2*node],self.tree[2*node+1])
    def _update(self,node,l,r,pos,val):
        while l!=r:
            mid = (l+r)>>1
            if pos>=mid+1:
                l = mid+1
                node = 2*node+1
            else:
                node *= 2
                r = mid
        self.tree[node] = Node(val,val,val,val)
        while node>1:
            node >>= 1
            self.tree[node] = SegmentTree.f(self.tree[2*node],self.tree[2*node+1])
    def update(self,pos,val):
        self._update(1,0,self.n-1,pos,val)

def solve():
    n,q = LII()
    L = LII()
    seg = SegmentTree(L)
    for _ in range(q):
        ind,r = LII()
        seg.update(ind-1,r)
        print(max(0,seg.tree[1].maxi))

    #L1 = LII()
    #st = SI()
solve()

File: Range Queries/Subarray_Sum_Queries.cpp
// Choose cpp when the time
// limit will for sure give tle in pypy
#include <bits/stdc++.h>
#include <ext/pb_ds/assoc_container.hpp>
#include <ext/pb_ds/tree_policy.hpp>
using namespace __gnu_pbds;
using namespace std;

#define INF LLONG_MAX
#define fastio ios::sync_with_stdio(false); cin.tie(nullptr); cout.tie(nullptr);
#define len(arr) arr.size()
#define f first
#define s second
#define pb push_back
#define all(x) x.begin(), x.end()
#define range(i, n) for (int i = 0; i < (n); ++i)
#define rangea(i, a, b) for (int i = (a); i < (b); ++i)
#define int long long
#define MOD1 998244353
#define MOD 1000000007
using vi = vector<int>;
using vii = vector<vi>;
using viii = vector<vii>;
using ll = long long;
using pi = pair<int,int>;
#pragma GCC optimize("O3,unroll-loops")
#pragma GCC target("avx2,bmi,bmi2,lzcnt,popcnt")
mt19937 rng(chrono::steady_clock::now().time_since_epoch().count());

#ifdef LOCAL
#define dbg(x...) cout<<"["<<#x<<"] = "; print(x);
#else
#define dbg(x...)
#endif

template <typename T>
using ordered_multiset = tree<
    std::pair<T, int>,
    null_type,
    std::less<std::pair<T, int>>,
    rb_tree_tag,
    tree_order_statistics_node_update
>;
template<class T> bool chmin(T& a,const T& b){ if(b<a){ a=b; return 1;} return 0; }
template<class T> bool chmax(T& a,const T& b){ if(a<b){ a=b; return 1;} return 0;}
template<class T> using minpq = priority_queue<T,vector<T>,greater<T>>;
template<class T> using maxpq = priority_queue<T>;
template<typename T> istream& operator>>(istream& in, vector<T>& v) {for (auto& x : v) in >> x;return in;}
template<typename T1, typename T2> istream& operator>>(istream& in, pair<T1, T2>& p) {in >> p.first >> p.second;return in;}
template<class T> void _print(const T &x){ cout << x; }
template<class T,class U> void _print(const pair<T,U> &p){cout<<"(";_print(p.first);cout<<",";_print(p.second);cout<<")";}
template<class T>
void _print(const vector<T> &v){
    cout<<"[";bool f=0;
    for(const auto &x: v){
        if(f) cout<<", ";
        _print(x);
        f=1;
    }
    cout<<"]";
}
template<class T>
void _print(const set<T> &v){
    cout<<"{";
    bool f=0;
    for(const auto &x: v){
        if(f) cout<<", ";
        _print(x);
        f=1;
    }
    cout<<"}";
}
template<class T>
void _print(const unordered_set<T> &v){
    cout<<"{";
    bool f=0;
    for(const auto &x: v){
        if(f) cout<<", ";
        _print(x);
        f=1;
    }
    cout<<"}";
}
template<class T,class U>
void _print(const unordered_map<T,U> &m){
    cout<<"{";
    bool f=0;
    for(const auto &p: m){
        if(f) cout<<", ";
        _print(p);
        f=1;
    }
    cout<<"}";
}
template<class T,class U>
void _print(const map<T,U> &m){
    cout<<"{";
    bool f=0;
    for(const auto &p: m){
        if(f) cout<<", ";
        _print(p);
        f=1;
    }
    cout<<"}";
}
void print(){ cout<<"\n"; }
template<class T,class... Args>
void print(const T& a,const Args&... args){
    _print(a);
    if constexpr(sizeof...(args)) cout<<" ";
    print(args...);
}

int gcd(int a, int b) {if (b == 0) return a;return gcd(b, a % b);}
int lcm(int a, int b) {return a / gcd(a, b) * b;}
int II(){int a;cin>>a;return a;}
string SI(){string s;cin>>s;return s;}
vi LII(int n){vi a(n);cin>>a;return a;}
int rnd(int l,int r){return uniform_int_distribution<int>(l,r)(rng);}

// String hashing: sh/shclass, Number: numtheory, SparseTable: SparseTable, SortedList: sortl
// Segment Tree(lazy propogation): SegmentTree, Merge Sort Tree: sorttree, Trie/Treap: Tries
// binary indexed tree: BIT, Segment Tree(point updates): SegmentPoint, Convex Hull: hull
// Combinatorics: pnc, Diophantine Equations: dpheq, Graphs: graphs, Centroid Decomp.: graph_decom
// Persistent Segment Tree: perseg, FreqGraphs: bgraph, GrapthOth: graphoth, DSU: DSU, FFT:fft
//  Rollback/Par DSU: rbdsu, treap: treap
// Template : https://github.com/OmAmar106/Template-for-Competetive-Programming
using t2 = tuple<int,int,int,int>;
t2 f(t2 &node1,t2 &node2){
    return {max(max(get<0>(node1),get<0>(node2)),get<1>(node2)+get<2>(node1)),max(get<1>(node1),get<1>(node2)+get<3>(node1)),max(get<2>(node2),get<3>(node2)+get<2>(node1)),get<3>(node1)+get<3>(node2)};
}

class SegmentTree{
    public:
        int n;vector<tuple<int,int,int,int>> tree;
        SegmentTree(vi L){
            n = len(L);
            tree.resize(4*n);
            build(1,0,n-1,L);
        }
        void build(int node, int start, int end, vi &L){
            if(end<start){
                return;
            }
            if(start==end){
                // print(node,len(tree));
                tree[node] = {L[start],L[start],L[start],L[start]};
                return;
            }
            int mid = (start+end)>>1;
            build(2*node,start,mid,L);
            build(2*node+1,mid+1,end,L);
            tree[node] = f(tree[2*node],tree[2*node+1]);
        }
        void update(int node,int start,int end,int pos,int val){
            if(end<start){
                return;
            }
            if(start==end){
                tree[node] = {val,val,val,val};
                return;
            }
            int mid = (start+end)>>1;
            if(pos<=mid){update(2*node,start,mid,pos,val);}
            else{update(2*node+1,mid+1,end,pos,val);}
            tree[node] = f(tree[2*node],tree[2*node+1]);
        }
};  

void solve() {
    int n,q;
    cin>>n>>q;

    vi L(n);
    for(auto &i:L){
        cin>>i;
    }
    // print(L);
    SegmentTree seg = SegmentTree(L);

    while(q--){
        int ind,val;
        cin>>ind>>val;
        ind--;
        seg.update(1,0,n-1,ind,val);
        print(max(0ll,get<0>(seg.tree[1])));
    }

}

int32_t main() {
    fastio
    int t=1;
    // cin >> t;
    while (t--) solve();
    return 0;
}